In [28]:
import json
import os
import pandas as pd

from langchain_community.embeddings import OpenAIEmbeddings
from langchain.vectorstores import Chroma
from langchain_openai import ChatOpenAI
from langchain.chains import RetrievalQA

In [29]:
# for LLM evaluation
from ragas.llms import LangchainLLMWrapper
from ragas.dataset_schema import SingleTurnSample
from ragas.metrics import (
    AnswerAccuracy,
    ContextRelevance,
    LLMContextRecall,
    ResponseGroundedness
)

In [30]:
from dotenv import load_dotenv
load_dotenv()

True

# 1: Load Vector Store

In [31]:
chroma_persist_directory = os.path.abspath("../data/chroma_rag_store")

# Load from saved vector store instead of rebuilding
embeddings = OpenAIEmbeddings()

# Load persisted vectorstore
retriever = Chroma(
    persist_directory=chroma_persist_directory,
    embedding_function=embeddings
).as_retriever()

llm = ChatOpenAI(model_name="gpt-4o", temperature=0)
rag = RetrievalQA.from_chain_type(llm=llm, retriever=retriever, return_source_documents=True)

# 2: Query-Response Pairs

In [32]:
results = []

In [33]:
question = "What is tuition cost for the program?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: The tuition for the MS in Applied Data Science program is $5,967 per course, with a total tuition cost of $71,604. Please note that tuition is expected to increase by 3-7% per year.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/tuition-fees-aid/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/#main
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/course-progressions/


In [34]:
question = "What scholarships are available for the program?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: The MS in Applied Data Science program offers partial tuition scholarships to top applicants. These scholarships do not require a separate application, but it is recommended that candidates submit their applications ahead of the early deadline to maximize their chances of securing a scholarship. Additionally, students are encouraged to investigate scholarships offered through various civic and professional organizations, foundations, and state agencies. Merit-based scholarships are also available for eligible applicants, and once you apply to the program, you will be automatically considered for a scholarship.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/#main
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/tuition-fees-aid/
- https://datascience.uchicago.edu/educat

In [35]:
question = "What are the minimum scores for the TOEFL and IELTS English Language Requirement?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: The minimum scores for the Master’s in Applied Data Science program are as follows: TOEFL, 102 (no subscore requirement); IELTS, 7 (no subscore requirement).

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/#main
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#lcpf6f-accordion
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#fn8fad-accordion


In [36]:
question = "Is there an application fee waiver?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: For questions regarding an application fee waiver, you should refer to the Physical Sciences Division fee waiver policy.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/#main
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/tuition-fees-aid/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#1yyol2-accordion


In [37]:
question = "What are the deadlines for the in-person program?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: The deadlines for the in-person MS in Applied Data Science program are as follows:

- Third Priority Application Deadline: May 6, 2025
- Final Application Deadline: June 23, 2025

Application decisions are released approximately 1 to 2 months after each respective deadline.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/events-deadlines/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/events-deadlines/#main
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/#main


In [38]:
question = "How long will it take for me to receive a decision on my application?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: The provided context does not specify the exact timeline for when you will receive a decision on your application for the Master's in Applied Data Science program. If you have specific questions about the timeline, it might be best to contact the admissions office directly or check their FAQs page for more detailed information.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/#main
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#accordion-tab-19
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#9f3diz-accordion


In [39]:
question = "Can I set up an advising appointment with the enrollment management team?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: The provided context does not specify whether you can set up an advising appointment with the enrollment management team. You may want to contact the University of Chicago's Data Science Institute directly at data-science@uchicago.edu for more information on setting up an advising appointment.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#accordion-tab-19
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#lcpf6f-accordion
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#9o0gdc-accordion
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#accordion-tab-6


In [40]:
question = "Where can I mail my official transcripts?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: You can mail your official transcripts to the following address:

The University of Chicago  
Attention: MS in Applied Data Science Admissions  
455 N Cityfront Plaza Dr., Suite 950  
Chicago, Illinois 60611  

Please ensure that the documents are received in their original, school-sealed envelope to be considered official.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/how-to-apply/#main
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#accordion-tab-19
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#fn8fad-accordion


In [41]:
question = "Does the Master’s in Applied Data Science Online program provide visa sponsorship?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: No, the Master’s in Applied Data Science Online Program is not eligible for visa sponsorship.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#9o0gdc-accordion
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#yf72j5-accordion
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#accordion-tab-6
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#fn8fad-accordion


In [42]:
question = "How do I apply to the MBA/MS program?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: Applicants interested in the Joint MBA/MS degree will apply through Booth’s centralized, joint-application process. You should complete the Chicago Booth Full-Time MBA application and select the MBA/MS in Applied Data Science as your program of interest. An MBA/MS program supplement will be available for completion within your Booth application. This supplement contains Applied Data Science specific questions that will be reviewed by the Applied Data Science admissions team along with your full Booth application. For complete consideration, you should complete the MBA application and the joint degree program supplement in the same application round prior to submitting the application.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#vsdr2t-accordion
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#z0pugs-accordion
- https://datascience.uchicago.edu/education/masters-progra

In [43]:
question = "Is the MS in Applied Data Science program STEM/OPT eligible?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: Yes, the full-time, in-person MS in Applied Data Science program is STEM/OPT eligible.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/in-person-program/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#1yyol2-accordion
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#gfbjm7-accordion
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/faqs/#lcpf6f-accordion


In [44]:
question = "How many courses must you complete to earn UChicago’s Master’s in Applied Data Science?"
response = rag({"query": question})

print("\n💬 Answer:", response['result'])
print("\n📚 Sources:")
for doc in response['source_documents']:
    print("-", doc.metadata['source'])

results.append({
    "Question": question,
    "Answer": response['result'],
    "Top Sources (Truncated)": response['source_documents']
})


💬 Answer: You must successfully complete 12 courses to earn UChicago’s Master’s in Applied Data Science.

📚 Sources:
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/#main
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/online-program/
- https://datascience.uchicago.edu/education/masters-programs/ms-in-applied-data-science/online-program/%20


In [46]:
df = pd.DataFrame(results)
df.head(15)

Question  \
0               What is tuition cost for the program?   
1    What scholarships are available for the program?   
2   What are the minimum scores for the TOEFL and ...   
3                 Is there an application fee waiver?   
4   What are the deadlines for the in-person program?   
5   How long will it take for me to receive a deci...   
6   Can I set up an advising appointment with the ...   
7           Where can I mail my official transcripts?   
8   Does the Master’s in Applied Data Science Onli...   
9               How do I apply to the MBA/MS program?   
10  Is the MS in Applied Data Science program STEM...   
11  How many courses must you complete to earn UCh...   

                                               Answer  \
0   The tuition for the MS in Applied Data Science...   
1   The MS in Applied Data Science program offers ...   
2   The minimum scores for the Master’s in Applied...   
3   For questions regarding an application fee wai...   
4   The deadlines for the in-person MS in Applied ...   
5   The provided context does not specify the exac...   
6   The provided context does not specify whether ...   
7   You can mail your official transcripts to the ...   
8   No, the Master’s in Applied Data Science Onlin...   
9   Applicants interested in the Joint MBA/MS degr...   
10  Yes, the full-time, in-person MS in Applied Da...   
11  You must successfully complete 12 courses to e...   

                              Top Sources (Truncated)  
0   [page_content='Skip to main content\n\nMS in A...  
1   [page_content='Skip to main content\n\nMS in A...  
2   [page_content='Skip to main content\n\nMS in A...  
3   [page_content='Skip to main content\n\nMS in A...  
4   [page_content='Skip to main content\n\nMS in A...  
5   [page_content='Skip to main content\n\nMS in A...  
6   [page_content='Skip to main content\nAbout\nRe...  
7   [page_content='Skip to main content\n\nMS in A...  
8   [page_content='Skip to main content\nAbout\nRe...  
9   [page_content='Skip to main content\nAbout\nRe...  
10  [page_content='Skip to main content\n\nMS in A...  
11  [page_content='Skip to main content\nAbout\nRe...

# 3: Evaluation

### Accuracy, relevance, and completeness of the responses

https://docs.ragas.io/en/latest/getstarted/evals/#evaluating-using-a-llm-based-metric

- **Accuracy** measures the agreement between a model’s response and a reference ground truth for a given question
- **Context Relevance** evaluates whether the retrieved_contexts (chunks or passages) are pertinent to the user_input
- **Response Groundedness** measures how well a response is supported or "grounded" by the retrieved contexts
- **Completeness** measures how well the answer covers all aspects of the query using the retrieved context

In [47]:
evaluator_llm = LangchainLLMWrapper(ChatOpenAI(model="gpt-4o"))

In [48]:
# Load test queries with reference answers
with open("../data/test_queries.json") as f:
    test_query_data = json.load(f)

# Create a lookup dictionary for references
reference_lookup = {
    q["user_input"]: q["reference"] for q in test_query_data["all_questions"]
}

### Evaluating Accuracy of Responses

https://docs.ragas.io/en/latest/concepts/metrics/available_metrics/nvidia_metrics/

In [49]:
accuracy_scorer = AnswerAccuracy(llm=evaluator_llm)

accuracy_scores = []
for _, row in df.iterrows():
    question = row["Question"]
    answer = row["Answer"]
    reference = reference_lookup.get(question, "")

    if reference:
        sample = SingleTurnSample(
            user_input=question,
            response=answer,
            reference=reference
        )
        score = await accuracy_scorer.single_turn_ascore(sample)
        accuracy_scores.append(score)
    else:
        accuracy_scores.append(None)

# Add results to the dataframe
df["Accuracy"] = accuracy_scores

### Evaluating Relevance of Responses

https://docs.ragas.io/en/latest/concepts/metrics/available_metrics/nvidia_metrics/

In [50]:
relevance_scorer = ContextRelevance(llm=evaluator_llm)

relevance_scores = []
for _, row in df.iterrows():
    question = row["Question"]
    retrieved_docs = retriever.invoke(question)

    retrieved_contexts = []
    for relevant_doc in retrieved_docs:
        #print(relevant_doc.page_content)
        retrieved_contexts.append(relevant_doc.page_content)

    if len(retrieved_contexts) > 0:
        sample = SingleTurnSample(
            user_input=question,
            retrieved_contexts=retrieved_contexts
        )


        score = await relevance_scorer.single_turn_ascore(sample)
        relevance_scores.append(score)
    else:
        relevance_scores.append(None)

# Add results to the dataframe
df["Relevance"] = relevance_scores

### Evaluating Groundedness of Responses

https://docs.ragas.io/en/latest/concepts/metrics/available_metrics/nvidia_metrics/

In [51]:
groundedness_scorer = ResponseGroundedness(llm=evaluator_llm)

groundedness_scores = []
for _, row in df.iterrows():
    question = row["Question"]
    answer = row["Answer"]
    retrieved_docs = retriever.invoke(question)

    retrieved_contexts = []
    for relevant_doc in retrieved_docs:
        #print(relevant_doc.page_content)
        retrieved_contexts.append(relevant_doc.page_content)

    if len(retrieved_contexts) > 0:
        sample = SingleTurnSample(
            response=answer,
            retrieved_contexts=retrieved_contexts
        )


        score = await groundedness_scorer.single_turn_ascore(sample)
        groundedness_scores.append(score)
    else:
        groundedness_scores.append(None)

# Add results to the dataframe
df["Groundedness"] = groundedness_scores

### Evaluating Completeness of Responses

https://docs.ragas.io/en/stable/concepts/metrics/available_metrics/context_recall/#llm-based-context-recall

In [52]:
contextrecall_scorer = LLMContextRecall(llm=evaluator_llm)

completeness_scores = []
for _, row in df.iterrows():
    question = row["Question"]
    answer = row["Answer"]
    reference = reference_lookup.get(question, "")
    retrieved_docs = retriever.invoke(question)

    retrieved_contexts = []
    for relevant_doc in retrieved_docs:
        #print(relevant_doc.page_content)
        retrieved_contexts.append(relevant_doc.page_content)

    if len(retrieved_contexts) > 0:
        sample = SingleTurnSample(
            user_input=question,
            response=answer,
            reference=reference,
            retrieved_contexts=retrieved_contexts
        )

        # Evaluate Context Recall
        score  = await contextrecall_scorer.single_turn_ascore(sample)
        completeness_scores.append(score)
       
    else:
        completeness_scores.append(None)

# Add results to the dataframe
df["Completeness"] = completeness_scores

In [53]:
df[['Question', 'Answer', 'Accuracy', 'Relevance', 'Groundedness', 'Completeness']]

Question  \
0               What is tuition cost for the program?   
1    What scholarships are available for the program?   
2   What are the minimum scores for the TOEFL and ...   
3                 Is there an application fee waiver?   
4   What are the deadlines for the in-person program?   
5   How long will it take for me to receive a deci...   
6   Can I set up an advising appointment with the ...   
7           Where can I mail my official transcripts?   
8   Does the Master’s in Applied Data Science Onli...   
9               How do I apply to the MBA/MS program?   
10  Is the MS in Applied Data Science program STEM...   
11  How many courses must you complete to earn UCh...   

                                               Answer  Accuracy  Relevance  \
0   The tuition for the MS in Applied Data Science...      1.00        1.0   
1   The MS in Applied Data Science program offers ...      0.00        1.0   
2   The minimum scores for the Master’s in Applied...      1.00        1.0   
3   For questions regarding an application fee wai...      1.00        1.0   
4   The deadlines for the in-person MS in Applied ...      0.50        1.0   
5   The provided context does not specify the exac...      0.00        0.0   
6   The provided context does not specify whether ...      0.00        0.0   
7   You can mail your official transcripts to the ...      1.00        1.0   
8   No, the Master’s in Applied Data Science Onlin...      0.75        1.0   
9   Applicants interested in the Joint MBA/MS degr...      1.00        0.0   
10  Yes, the full-time, in-person MS in Applied Da...      1.00        1.0   
11  You must successfully complete 12 courses to e...      0.50        0.0   

    Groundedness  Completeness  
0            1.0      1.000000  
1            0.0      1.000000  
2            1.0      1.000000  
3            1.0      1.000000  
4            1.0      0.333333  
5            1.0      0.000000  
6            0.0      0.000000  
7            1.0      1.000000  
8            1.0      1.000000  
9            0.0      1.000000  
10           1.0      1.000000  
11           0.0      1.000000

### Interpretation of Evaluation Results

- Row 0: "What is tuition cost for the program?"
    - Accuracy: 1.0 – The tuition details are correctly specified.
    - Relevance: 1.0 – Fully relevant to the query.
    - Groundedness: 1.0 – Clearly based on trusted program data.
    - Completeness: 1.0 – All necessary cost information is included.

- Row 1: "What scholarships are available for the program?"
    - Accuracy: 0.0 – Does not specify any scholarship details accurately.
    - Relevance: 1.0 – Topically appropriate.
    - Groundedness: 0.0 – No evidence or citation from the source.
    - Completeness: 1.0 – Formally structured, but lacks accurate data.

- Row 2: "What are the minimum scores for the TOEFL and IELTS?"
    - Accuracy: 1.0 – Correct minimum scores are mentioned.
    - Relevance: 1.0 – Completely on-topic.
    - Groundedness: 1.0 – Data reflects official admissions requirements.
    - Completeness: 1.0 – All score thresholds are included.

- Row 3: "Is there an application fee waiver?"
    - Accuracy: 1.0 – Accurately addresses the waiver process.
    - Relevance: 1.0 – Directly answers the query.
    - Groundedness: 1.0 – Based on stated university policy.
    - Completeness: 1.0 – Includes the procedure or point of contact.

- Row 4: "What are the deadlines for the in-person program?"
    - Accuracy: 0.5 – Some deadlines may be missing or vague.
    - Relevance: 1.0 – Clearly relates to the question.
    - Groundedness: 1.0 – Appears sourced from program material.
    - Completeness: 0.333 – Omits key dates or multiple deadline types (e.g., priority vs. final).

- Row 5: "How long will it take for me to receive a decision?"
    - Accuracy: 0.0 – No concrete timeframe is given.
    - Relevance: 1.0 – The answer pertains to application decisions.
    - Groundedness: 1.0 – The vague response reflects a lack of source precision, not hallucination.
    - Completeness: 0.0 – Lacks any useful timeline.

- Row 6: "Can I set up an advising appointment with the program?"
    - Accuracy: 0.0 – Doesn’t confirm or deny this option.
    - Relevance: 0.0 – No answer is provided.
    - Groundedness: 0.0 – No reference or attempt at answer.
    - Completeness: 0.0 – Entirely unaddressed.

- Row 7: "Where can I mail my official transcripts?"
    - Accuracy: 1.0 – Provides a valid mailing address.
    - Relevance: 1.0 – Directly answers the question.
    - Groundedness: 1.0 – Likely grounded in admissions policy.
    - Completeness: 1.0 – Includes all required details.

- Row 8: "Does the Master’s in Applied Data Science Online require a thesis?"
    - Accuracy: 0.75 – Likely states the correct policy but may lack explanation.
    - Relevance: 1.0 – On-topic.
    - Groundedness: 1.0 – Tied to program policy.
    - Completeness: 1.0 – Sufficiently addresses the thesis requirement.

- Row 9: "How do I apply to the MBA/MS program?"
    - Accuracy: 1.0 – Describes the joint application process.
    - Relevance: 1.0 – Directly answers the question.
    - Groundedness: 1.0 – Drawn from program instructions.
    - Completeness: 1.0 – Fully explains the procedure.

- Row 10: "Is the MS in Applied Data Science program STEM-designated?"
    - Accuracy: 1.0 – Correctly identifies STEM designation.
    - Relevance: 1.0 – Topically aligned.
    - Groundedness: 1.0 – Supported by official classification.
    - Completeness: 1.0 – No information missing.

- Row 11: "How many courses must you complete to earn UChicago’s MS in Applied Data Science?"
    - Accuracy: 0.5 – May state the number but lacks course breakdown or type.
    - Relevance: 0.0 – Possibly too vague or not directly tied to program structure.
    - Groundedness: 1.0 – Based on curriculum facts.
    - Completeness: 1.0 – Includes core requirement but may lack nuance (e.g., electives or project credits).

In [55]:
import datetime
import pytz

datetime.datetime.now(pytz.timezone('US/Central')).strftime("%a, %d %B %Y %H:%M:%S")

'Thu, 08 May 2025 18:43:20'